Safe arbitrary code execution

TheConner opened the pull request:
This comes in two parts:
1. First we migrated to use Serenity + Poise, as we need a way to dispatch Modal interaction responses. Slash commands sadly have no facilities for code block inputs (or even multiline inputs), so we have to use modals
2. Implementing the actual feature, echoing part of commit message below as it summarizes the changes

**Summary of changes**
- Poise framework (needed for modals / multiline input under command framework)
- Better configuration framework (containers need a bunch of config items set through environment variables)
- Container management (ability to pull, work with, and execute containers)
- Docker-in-docker support (we use podman but it should be fine...)
- Code execution bot command (depends on above)

This is a port of [RustBot](https://github.com/TheConner/RustBot) -- although, this feature does *depend* on rustbot as we are leveraging the [rustbot-runner container artifact from my repo](https://github.com/TheConner/RustBot/pkgs/container/rustbot-runner)

**Other notes**
- I changed how bot registration works as part of the poise refactor. Run !register to set up the bot commands initially (use for guild only option when under development). In the future we should limit this to only admins / bot developers.
- Runnable trait allows any string to be executed in container. Just call `.run()` on any string, then you get to play with the output from Stdout/Stderr
- Existing question/quiz functionality refactored -- slight changes made to pass through Serenity structs from Poise

TheConner committed:
Migrate to poise framework

In order to implement code injection, we need to use poise framework which works ontop of serenity. Poise provides a bot command framework that allows us to use modals and other multiline inputs

- Ported existing quiz functionaly to poise library (just using serenity structs inside ApplicationContext)
- Refactored commands to be in a `/commands/` directory vs a single handler file

TheConner committed:
Rust code execution feature

This commit is a large refactor to the repo to support
- Poise framework (needed for modals / multiline input under command framework)
- Better configuration framework (containers need a bunch of config items set through environment variables)
- Container management (ability to pull, work with, and execute containers)
- Docker-in-docker support (we use podman but it should be fine...)
- Code execution bot command (depends on above)

This is a port of RustBot https://github.com/TheConner/RustBot -- although, this feature does *depend* on rustbot as we are leveraging the rustbot-runner container artifact from my repo https://github.com/TheConner/RustBot/pkgs/container/rustbot-runner

Other notes:

- I changed how bot registration works as part of the poise refactor. Run !register to set up the bot commands initially (use for guild only option when under development)
- Runnable trait allows any string to be executed in container.
- Existing question/quiz functionality refactored -- slight changes made to pass through Serenity structs from Poise

TheConner committed:
Satisfy fmt and Clippy

AngelOnFira commented on src/commands/run.rs:
```diff
@@ -0,0 +1,163 @@
+#[allow(dead_code)] // fields only used for Debug print
+struct RunModal {
+    #[name = "Code you want to run"]
+    #[placeholder = "fn main() {\n    println!(\"Hello, world!\");\n}"]
```
I'm curious about what might work best for editing a ~20-line function. I think the code that is part of the question would be a good placeholder? It would be really cool if a modal could feel like a mini editor.

TheConner replied:
Good point... A few things would help with editing a 20 line function
- Having monospace support would be great, but unfortunately we can't get that with modals or slash commands just yet. The original RustBot had this feature by *not* using the command framework and just having users enter code blocks in markdown, but this is kind of a waste of time here as Discord is canning the "classic" functionality in favor of the command framework (unless you get approval from them if you *need* that functionality)
- You could inject the question code sample as the `value` of the modal instead of the placeholder, like so https://docs.rs/poise/0.2.1/src/poise/modal.rs.html#149

AngelOnFira commented on src/configuration.rs:
```diff
@@ -0,0 +1,42 @@
+/// Sets the maximum amount of virtual CPUs available to the child container
+pub const CONTAINER_CPU: &ConfigurableItem<&str> = &ConfigurableItem {
+    environment_variable: "CONTAINER_CPU",
+    default_value: "0.5", // you get 1/2 of a cpu, i'm being generous
```
:laughing:

TheConner replied:
Something worth noting here: if you're running using docker-in-docker these limits won't apply (see [here](https://github.com/Summer-of-Rust/ferris-bot/pull/6/commits/26ee1ff71419155b67afa87c6a7bcae2984580c3#diff-bac1ed91dd1c520d98b8bb94164bac1240b9d076e21cfd295c50f374c8e7c7d5R33)) -- resource limitations would need to be applied to the outer container. 

(of course, if there's a workaround for this now this limitation can be resolved)

AngelOnFira commented:
Alright, I think this is looking pretty good. I'll give it a test tomorrow, and hopefully be able to try it out at this week's sessions :)

TheConner committed:
Code execution improvements and bug fixes

- Code syntax formatting
- Using message embeds to show code, output, and error
- Support for showing stdout and stderr concurrently
- Better output truncating

TheConner committed:
Satisfy fmt and clippy

TheConner committed:
Fix bug: response code is double formattted

AngelOnFira committed:
Changed fields to be different rows in embeds
